### ISSUE TITLE
abandoned or feature complete/support-only mode?

### ISSUE THREAD

**Issue by ThePiGrepper** [michaelforney/wld#13]
hi, I've been trying to find a simple,small wayland-client library to use for some ports from simple X programs(some suckless ones likes the ones you have already forks of. ie: st,dmenu)

I was wondering what are your plans/ideas regarding this project in the near future? I ask because as I said I've been searching for a cairo+pango alternative, a simple non-bloated library to use for programs which dont require much else besides rendering some text and some primitives, and this project is one of the few I've found that actually fits that description quite nicely.

I've been studying the source code and the lack of documentation sometimes makes it a bit harder. I understand that it probably wasnt your original intent that ppl besides yourself start using it. However, I'd like to know what are your plans regarding this library, so that maybe I can help you with documentation writing, maybe writing a summary so that the goals of the project are clear to anyone who encounters this site, and ofc submitting patches and bug fixing.
I'd like for this to be a real alternative to other bigger libraries, which have a very different usecase. I'd like to know your opinion on this.
PS: sorry if Im writing this as an 'issue', I probably should've send an email...

**Comment by markand** [comment]
Can someone ELI5 why the library has explicit support for graphics card? I mean, isn't wayland meant to be used directly with higher level primitives such as OpenGL?

**Comment by michaelforney** [comment]
> Can someone ELI5 why the library has explicit support for graphics card?

There are a few reasons:

- Mesa requires dynamically loaded modules, and can't be linked statically (which is needed for oasis, where everything is statically linked).
- Mesa uses C++ for substantial parts, and C++ does not permit a simple or light-weight implementation (oasis does not use any software written in C++, and aims to build the whole system with [cproc](https://man.sr.ht/~mcf/cproc/doc/software.md#oasis)).
- I'm fairly certain that even in the C parts of Mesa, many GNU C extensions are used.
- In some cases, you even need to link LLVM to compile shaders.

All of this seems pretty excessive to me since all I want to do is use my graphics card to fill rectangles with a solid color, or copy a region from one buffer to another.

Of course, I'm not opposed to having a wld backend that renders using OpenGL or Vulkan so that it works for all hardware supported by mesa (for people who don't care about the above). In fact, in [libblit](https://git.sr.ht/~mcf/libblit), the WIP successor project mentioned above, I am doing just that. But, having alternative backends that work with just libdrm will remain a requirement for me (at least, for the hardware I have).

> I mean, isn't wayland meant to be used directly with higher level primitives such as OpenGL?

Actually, wayland only includes one interface for creating a `wl_buffer` in the core protocol, and that is `wl_shm`. It is not tied to any particular mechanism of rendering.

But, wld does use the same `wayland_drm.xml` protocol as Mesa, so clients rendered with wld will work fine with a compositor using OpenGL, and a compositor written with wld can display surfaces from clients using OpenGL.
